src/libexpr/eval-inline.hh

```diff
@@ -84,9 +88,33 @@ Env & EvalState::allocEnv(size_t size)
 [[gnu::always_inline]]
 void EvalState::forceValue(Value & v, const PosIdx pos)
 {
+    auto type = v.internalType.load(std::memory_order_acquire);
+
+    if (isFinished(type))
+        goto done;
+
+    if (type == tThunk) {
+        try {
+            if (!v.internalType.compare_exchange_strong(type, tPending, std::memory_order_acquire, std::memory_order_acquire)) {
+                if (type == tPending || type == tAwaited) {
+                    waitOnThunk(v, type == tAwaited);
+                    goto done;
+                }
+                if (isFinished(type))
+                    goto done;
+                printError("NO LONGER THUNK %x %d", this, type);
+                abort();
+            }
+            Env * env = v.payload.thunk.env;
+            Expr * expr = v.payload.thunk.expr;
+            expr->eval(*this, *env, v);
+        } catch (...) {
+            v.mkFailed();
+            throw;
+        }
+    }
+    #if 0
```

Comment:
I might be a bit rusty on this, but GHC black holes store a thread id (thread pointer?) in its blackholes.
Same thread => definitely infinite recursion
Different thread => check that the other thread is not (transitively) waiting for us. (iirc it might perform the same computation - which is why `unsafeDupablePerformIO` exists and is cheap). Nix could just block instead, after saving an edge in the thread waiting map that we used for checking.